### PR TITLE
(SERVER-2213) Upgrade Jetty to latest

### PR DESCRIPTION
**Pull request by justinstoller**

### CODE REVIEW

File: test/clj/puppetlabs/trapperkeeper/services/webserver/jetty9_default_config_test.clj [hunk 1]
```diff
@@ -88,19 +88,26 @@ react accordingly."
...
-(def selector-thread-count
-  "The number of selector threads that should be allocated per connector.  See:
-   https://github.com/eclipse/jetty.project/blob/jetty-9.4.1.v20170120/jetty-server/src/main/java/org/eclipse/jetty/server/ServerConnector.java#L223
-   and
-   https://github.com/eclipse/jetty.project/blob/jetty-9.4.1.v20170120/jetty-server/src/main/java/org/eclipse/jetty/server/Server.java#L403-L408
-   The number of selectors is twice the number of selector threads."
-  (* 2 (max 1 (min 4 (int (/ (ks/num-cpus) 2))))))
+(defn selector-thread-count
+  [max-threads]
+  "The number of selectors and selecto threads that should be allocated per
```
Comment (justinstoller) [inline test/clj/puppetlabs/trapperkeeper/services/webserver/jetty9_default_config_test.clj]: its actually just the threads, "selecto" threads

File: test/clj/puppetlabs/trapperkeeper/services/webserver/jetty9_default_config_test.clj [hunk 2]
```diff
@@ -88,19 +88,26 @@ react accordingly."
...
+       (min (int (/ max-threads 16))
+            (int (/ (ks/num-cpus) 2)))))
 
 (def acceptor-thread-count
   "The number of acceptor threads that should be allocated per connector.  See:
-   https://github.com/eclipse/jetty.project/blob/jetty-9.4.1.v20170120/jetty-server/src/main/java/org/eclipse/jetty/server/AbstractConnector.java#L194"
+   https://github.com/eclipse/jetty.project/blob/jetty-9.4.11.v20180605/jetty-server/src/main/java/org/eclipse/jetty/server/AbstractConnector.java#L202"
   (max 1 (min 4 (int (/ (ks/num-cpus) 8)))))
 
+(defn reserved-thread-count
```
Comment (justinstoller) [inline test/clj/puppetlabs/trapperkeeper/services/webserver/jetty9_default_config_test.clj]: Should I write about ReservedThreadPool here?

File: test/clj/puppetlabs/trapperkeeper/services/webserver/jetty9_default_config_test.clj [hunk 3]
```diff
@@ -143,15 +150,22 @@ react accordingly."
...
 (defn calculate-minimum-required-threads
-  "Calculate the minimum number threads that a single Jetty Server instance
-  needs.  See: https://github.com/eclipse/jetty.project/blob/jetty-9.4.1.v20170120/jetty-server/src/main/java/org/eclipse/jetty/server/Server.java#L384-L414"
-  [connectors]
-  (+ 1 (* connectors threads-per-connector)))
+  "Jetty calculates the minimum number of required threads based on the the
+  size of the machine it runs on and the number of threads in its thread pool.
+  The smaller we shrink the thread pool the fewer threads it will request.
+  This logic doesn't scale as the thread pool grows beyond 10 threads but
+  allows us to come up with a number that works for the max threads that isn't
```
Comment (justinstoller) [inline test/clj/puppetlabs/trapperkeeper/services/webserver/jetty9_default_config_test.clj]: s/the max threads/max threads/

File: test/clj/puppetlabs/trapperkeeper/services/webserver/jetty9_default_config_test.clj [hunk 4]
```diff
@@ -143,15 +150,22 @@ react accordingly."
...
         "Unexpected default for 'queue-max-size'")))
 
-(def threads-per-connector
+(defn required-threads-for-sized-threadpool
+  [threadpool-size]
   "The total number of threads needed per attached connector."
-  (+ acceptor-thread-count selector-thread-count))
+  (+ (reserved-thread-count threadpool-size)
+     (selector-thread-count threadpool-size)
+     acceptor-thread-count))
```
Comment (justinstoller) [inline test/clj/puppetlabs/trapperkeeper/services/webserver/jetty9_default_config_test.clj]: This doesn't seem to hold when the threadpool is super small and I don't know why.

**Comment by justinstoller** [comment]
Still need to at least update the actual documentation and test this in a puppetserver build.

**Comment by justinstoller** [comment]
@puppetlabs/puppetserver-maintainers I think this is ready to go. (@Sharpie you might be iterested in this as well).

**Comment by justinstoller** [comment]
@samwoods1 you and the SLV team might also be interested in this. We're hoping to upgrade Jetty in Puppet 6/Johnson and the threading defaults have changed. I don't think there's any reason for concern when upgrading, but those are famous last words.  (:

**Comment by justinstoller** [comment]
The travis build failed with something I don't see locally, re-kicking to see if it's transient.

**Comment by justinstoller** [comment]
@gguillotte This has changes that affect the defaults Jetty uses for its thread allocation. I've updated the documentation in this repo but I didn't see anything on the main site regarding "tuning jetty threads", afaict it's something the CS may do at engagements but isn't something we give general guidance on. Wanted to call this out if you know of any place we do so.

File: test/clj/puppetlabs/trapperkeeper/services/webserver/jetty9_default_config_test.clj (hunk 3 above)
Comment (Magisus) [reply test/clj/puppetlabs/trapperkeeper/services/webserver/jetty9_default_config_test.clj]: Are we worried about it not scaling? If not, why not?

**Comment by Magisus** [comment]
In the abstract this looks good to me, but I am curious about the implications of the threading changes, and I think it would be good if someone from CS took a look.

File: test/clj/puppetlabs/trapperkeeper/services/webserver/jetty9_default_config_test.clj (hunk 2 above)
Comment (amysahli) [reply test/clj/puppetlabs/trapperkeeper/services/webserver/jetty9_default_config_test.clj]: sure!

Comment (justinstoller) [reply test/clj/puppetlabs/trapperkeeper/services/webserver/jetty9_default_config_test.clj]: what do you think of the content now (with a quick blurb and a link)?

Comment (Magisus) [reply test/clj/puppetlabs/trapperkeeper/services/webserver/jetty9_default_config_test.clj]: That works for me, it's in line with the others.

**Pull request merged**